front/components/agent_builder/capabilities/knowledge/KnowledgeConfigurationSheet.tsx

```diff
@@ -26,147 +28,198 @@ import {
 import type {
   CapabilityFormData,
   ConfigurationSheetPageId,
   KnowledgeServerName,
 } from "@app/components/agent_builder/types";
 import type { AgentBuilderAction } from "@app/components/agent_builder/types";
 import {
+  ACTION_TYPE_TO_MCP_SERVER_MAP,
   capabilityFormSchema,
   CONFIGURATION_SHEET_PAGE_IDS,
+  isKnowledgeServerName,
+  isSupportedAgentBuilderAction,
 } from "@app/components/agent_builder/types";
 import { useSpacesContext } from "@app/components/assistant_builder/contexts/SpacesContext";
 import { DataSourceBuilderSelector } from "@app/components/data_source_view/DataSourceBuilderSelector";
 import type { DataSourceViewSelectionConfigurations } from "@app/types";
 
 interface KnowledgeConfigurationSheetProps {
-  capability: KnowledgeServerName | null;
+  capability?: KnowledgeServerName | null;
   onSave: (action: AgentBuilderAction) => void;
-  isOpen: boolean;
   onClose: () => void;
+  onOpen?: () => void;
   action?: AgentBuilderAction;
+  open: boolean;
 }
 
 export function KnowledgeConfigurationSheet({
   capability,
   onSave,
-  isOpen,
   onClose,
+  onOpen,
   action,
+  open,
 }: KnowledgeConfigurationSheetProps) {
-  // We store as state to control the timing to update the content for exit animation.
-  const [config, setConfig] = useState<CapabilityConfig | null>(null);
+  // Determine the config based on capability or action type
+  const config = useMemo(() => {
+    if (capability) {
+      return CAPABILITY_CONFIGS[capability];
+    }
+    if (action && isSupportedAgentBuilderAction(action)) {
+      const serverName = ACTION_TYPE_TO_MCP_SERVER_MAP[action.type];
+      return serverName && isKnowledgeServerName(serverName)
+        ? CAPABILITY_CONFIGS[serverName]
+        : null;
+    }
+    return null;
+  }, [capability, action]);
 
   const handleClose = () => {
     onClose();
-
-    // TODO: This is a hack and we should find a proper solution.
-    // Wait until closing animation ends, otherwise exit animation won't work.
-    setTimeout(() => {
-      setConfig(null);
-    }, 200);
   };
 
   const handleSave = (
     formData: CapabilityFormData,
-    dataSourceConfigurations: DataSourceViewSelectionConfigurations
+    dataSourceConfigurations: DataSourceViewSelectionConfigurations,
+    configToUse: CapabilityConfig
   ) => {
-    if (!config) {
-      // never going here
-      return;
-    }
-
     const newAction = generateActionFromFormData({
-      config,
+      config: configToUse,
       formData,
       dataSourceConfigurations,
       actionId: action?.id,
     });
 
     if (newAction) {
       onSave(newAction);
+      handleClose();
     }
+  };
+
+  const { control, handleSubmit } = useMemo(() => {
+    const dataSourceTree = action
+      ? getDataSourceTree(action)
+      : {
+          in: [],
+          notIn: [],
+        };
 
-    handleClose();
+    return createFormControl<CapabilityFormData>({
```

Comment:
I didn't dig much how it works but we need to reset selected spaces when the sheet is closed? Right now if you close you see the same space being selected